ow_power_system/src/power_system_node.cpp

```diff
@@ -14,91 +14,131 @@
-  std_msgs::Float64 soc_msg;
-  soc_msg.data = 0.0;
-  std_msgs::Float64 tempbat_msg;
-  tempbat_msg.data = 0.0;
-  int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
-  ROS_INFO ("Power system node running");
-
-  while (ros::ok()) {
 
+    // Voltage estimate based on pseudorandom noise and moving range
+    double voltage_est = min_voltage + static_cast <double> (rand()) / RAND_MAX * (max_voltage - min_voltage);
+
+    //individual msgs to be published
+    std_msgs::Int16 rul_msg;
+    rul_msg.data = 0;  // immediately set to a good default
+    std_msgs::Float64 soc_msg;
+    soc_msg.data = 0.0;
+    std_msgs::Float64 tempbat_msg;
+    tempbat_msg.data = 0.0;
+
+    int temp_index = 1; // Set to 1 for now (constant vector). This will change to median SOC or RUL index or fixed percentile
+
+    vector<map<MessageId, Datum<double>>> power_data; // GSAP Input Vector Definition
+    map<MessageId, Datum<double>> initialization_data; // GSAP Initialization Data
+    map<MessageId, Datum<double>> current_data; // GSAP Input Data at Current Timestep
+
+    // Set timestamp for initialization data to initialization time
+    init_power.setTime(init_time);
+    init_temp.setTime(init_time);
+    init_voltage.setTime(init_time);
+
+    // Determine timestamp for current callback
+    auto timestamp = system_clock::now();
+    
+    // Assign input parameters and corresponding timestamp
+    Datum<double> power = mech_power;
+    power.setTime(timestamp);
+    Datum<double> temperature = temp_est;
+    temperature.setTime(timestamp);
+    Datum<double> voltage = voltage_est;
+    voltage.setTime(timestamp);
+
+    // Populate initialization data structure
+    initialization_data.insert({MessageId::Watts, init_power});
+    initialization_data.insert({MessageId::Centigrade, init_temp});
+    initialization_data.insert({MessageId::Volts, init_voltage});
+    power_data.push_back(initialization_data);
+
+    // Populate current timestep data structure
+    current_data.insert({MessageId::Watts, power});
+    current_data.insert({MessageId::Centigrade, temperature});
+    current_data.insert({MessageId::Volts, voltage});
+    power_data.push_back(current_data);
+
+    // Update count and GSAP initialization parameters
+    update_count = update_count + 1.0;
+    init_power = mech_power;
+    init_temp = temp_est;
+    init_voltage = voltage_est;
+    init_time = timestamp;
+
+    // Create a configuration from a file
+    string config_path = ros::package::getPath("ow_power_system")+"/config/example.cfg";
+    ConfigMap config(config_path);
+    // Contruct a new prognoser using the prognoser factory. The prognoser
+    // will automatically construct an appropriate model, observer and predictor
+    // based on the values specified in the config.
+    auto prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", config);
```

Comment:
Why do you need to create and destroy a `prognoser` on each callback?!

Reply:
Addressed in latest commit.